File: algos/starter-algo/gamelib/unit.py
def is_stationary(unit_type, firewall_types):
    return unit_type in firewall_types

class GameUnit:
    """Holds information about a Unit. 

    Attributes:
        * unit_type (string): This unit's type
        * config (JSON): Contains information about the game
        * player_index (integer): The player that controls this unit. 0 for you, 1 for your opponent.
        * stability (integer): The health of the unit
        * x (integer): The x coordinate of the unit
        * y (integer): The y coordinate of the unit
        * stationary (bool): Whether or not this unit is a firewall
        * speed (float): A unit will move once every 1/speed frames
        * damage (int): The amount of damage this firwall unit will deal to enemy information.
        * damage_f (int): The amount of damage this information unit will deal to enemy firewalls.
        * damage_i (int): The amount of damage this information unit will deal to enemy information.
        * range (float): The effective range of this unit
        * stability (float): The current health of this unit
        * cost (int): The resource cost of this unit

    """
    def __init__(self, unit_type, config, player_index=None, stability=None, x=-1, y=-1):
        """ Initialize unit variables using args passed

        """
        self.unit_type = unit_type
        self.config = config
        self.player_index = player_index
        self.pending_removal = False
        self.x = x
        self.y = y
        self.__serialize_type()
        self.stability = self.max_stability if not stability else stability

    def __serialize_type(self):
        from .game_state import FIREWALL_TYPES, UNIT_TYPE_TO_INDEX, ENCRYPTOR
        self.stationary = is_stationary(self.unit_type, FIREWALL_TYPES)
        type_config = self.config["unitInformation"][UNIT_TYPE_TO_INDEX[self.unit_type]]
        if self.stationary:
            self.speed = 0
            if self.unit_type == ENCRYPTOR:
                self.damage = type_config["shieldAmount"]
            else:
                self.damage = type_config["damage"]
        else:
            self.speed = type_config["speed"]
            self.damage_f = type_config["damageF"]
            self.damage_i = type_config["damageI"]
        self.range = type_config["range"]
        self.max_stability = type_config["stability"]
        self.cost = type_config["cost"]

    def __toString(self):
        owner = "Friendly" if self.player_index == 0 else "Enemy"
        removal = ", pending removal" if self.pending_removal else ""
        return "{} {}, stability: {} location: {}{} ".format(owner, self.unit_type, self.stability, [self.x, self.y], removal)

    def __str__(self):
        return self.__toString()

    def __repr__(self):
        return self.__toString()

File: community/c++-algo/include/GameLib/src/unit.cpp
/*
Description: Implmentations for the unit.h header.
Last Modified: 07 Apr 2019
Author: Ryan Draves
*/

#include "unit.h"

namespace terminal {
    
    using std::string;
    using std::to_string;
    using json11::Json;

    /// Init the passed in params and pass off config initialization to serializeType()
    /// @param unitType Unit type of the game unit.
    /// @param config JSON game configuration passed in upon startup.
    /// @param stability Health of the game unit. 0 defaults it to maxStability.
    /// @param playerIndex 0 for ally unit (you), 1 for enemy unit.
    /// @param x X coordinate of the game unit's location.
    /// @param y Y coordinate of the game unit's location.
    GameUnit::GameUnit(UNIT_TYPE unitType, const Json &config, double stability, int playerIndex,
        int x, int y) : unitType(unitType), config(config), playerIndex(playerIndex),
        stability(stability), x(x), y(y), pendingRemoval(false) {
        serializeType();
        this->stability = stability == 0 ? maxStability : stability;
    }

    /// Finish the game unit initialization based on the config
    void GameUnit::serializeType() {
        stationary = isStationary(unitType);
        Json type_config = config["unitInformation"].array_items().at((int)unitType);
        if (stationary) {
            speed = 0;
            if (unitType == REMOVE) {
                pendingRemoval = true;
                // Initialize these in case someone decides to access them...
                damage = damageF = damageI = maxStability = range = cost = 0;
                return;
            }
            else if (unitType == ENCRYPTOR) {
                damage = type_config["shieldAmount"].int_value();
            }
            else {
                damage = type_config["damage"].int_value();
            }
            damageF = damage;
            damageI = damage;
        }
        else {
            speed = type_config["speed"].number_value();
            damageF = type_config["damageF"].int_value();
            damageI = type_config["damageI"].int_value();
        }
        range = type_config["range"].number_value();
        maxStability = type_config["stability"].number_value();
        cost = type_config["cost"].int_value();
    }

    /// Get a string representation of the game unit
    /// @return Elegant string representation of the game unit
    string GameUnit::toString() const {
        string owner = playerIndex == 0 ? "Friendly" : "Enemy";
        string removal = pendingRemoval ? ", pending removal" : "";
        return owner + " " + unitTypeStr(unitType) + ", stability: " +
            to_string(stability) + " location: [" + to_string(x) + ", " +
            to_string(y) + "]" + removal;
    }
}
